ophyd/v2/core.py

```diff
@@ -607,3 +616,78 @@ async def describe_configuration(self) -> Dict[str, Descriptor]:
 
     async def read_configuration(self) -> Dict[str, Reading]:
         return await merge_gathered_dicts(sig.read() for sig in self._conf_signals)
+
+
+KT = TypeVar("KT")
+
+
+class DeviceDict(MutableMapping[KT, Device]):
```

Comment:
Could this be written smaller if it subclassed `dict` and added a few methods? Also, if you subclass `Device` you don't need to specialcase the isinstance check in `get_device_children`:

```python
class DeviceDict(Dict[KT, Device], Device):
    def set_name(self, name: str):
        for child_name, device in self.items():
            device.set_name(f"{name}-{child_name}")
            device.parent = self

    async def connect(self, prefix: str, sim: bool):
        coros = {
            k: d.connect(sim)
            for k, d in self.items()
        }
        await wait_for_connection(**coros)
```

Reply:
interesting you wrote the connect method in this way; in the latest commit, the connect method just loops through all the keys in the dictionary and connects them in series. This is obviously not practical, but I found that if I used the method you've written above in the past this led to things not being connected properly.

I'm really not sure why exactly it's failing. But if you put that bit of code in and try to run the bottom test of test_core in ophyd/v2/tests it fails because the dict devices actually don't have their connect() methods run at all.

Reply:
note: have just checked and with the DeviceDict as above (except with `d.connect(prefix, sim)` instead) that test only works for this case:

```python
async def test_children_of_standard_readable_have_set_names_and_get_connected():
    parent = DummyStandardReadable("parent")
    parent.set_name("parent")
    assert parent.child1.name == "parent-child1"
    assert parent.dict_with_children[123].name == "parent-dict_with_children-123"
    assert parent.dict_with_children["abc"].name == "parent-dict_with_children-abc"

    await parent.connect()

    await parent.dict_with_children.connect(parent._init_prefix, False) #have to connect this otherwise test fails.

    assert parent.child1.connected
    assert parent.dict_with_children[123].connected
    assert parent.dict_with_children["abc"].connected

```

Reply:
In the latest commit I have proposed a solution to this problem... but I don't think it's pretty.